### DISCUSSION TITLE
UWLSWD validation

### DISCUSSION THREAD

**Discussion by gerontakos** [uwlib-cams/uwlswd#66] [technical]
We can start by writing something that validates the RDF/XML, which is the file we create manually at the outset of creating a dataset.

@gerontakos should create an XML Schema to validate RDF/XML, at least for instances of skos:ConceptScheme.

**Comment by briesenberg07** [comment]
+1  
More ideas:
- Use rdflib to compare an old graph (rdf/xml or other serialization prior to running main.py) with a new graph (rdf/xml or other serialization after running main.py) in a meaningful way. I currently just count triples (`len(g)`) but I'm sure rdflib would allow for a much more meaningful check!

**Comment by gerontakos** [comment]
That's a good idea and shows we need validation/testing at many points of the process. I was just trying to catch things like dct:titl  or skos:Croncept, or "no literal value present for the required dc:creator property" (when the value is absent, currently, main.py throws an error to standard output and to the output file, I believe; but if it dc:creator present and does not have a literal value, I don't know what happens -- it doesn't matter, that specific problem is not the issue here). So I was just trying to catch human error (most likely) while manually creating an RDF/XML main document. Benjamin's validation is to check on the output of main.py, which, I believe, is more important and should come first. The cost of human error should be small; the cost of errors in our processing could be enormous.

**Comment by briesenberg07** [comment]
> trying to catch things like dct:titl or skos:Croncept

A couple of related issues in this category:
- https://github.com/uwlib-cams/uwlswd_vocabs/issues/10
- https://github.com/uwlib-cams/uwlswd_vocabs/issues/9

**Comment by gerontakos** [comment]
BTW I checked the results of main.py, on my first try, only by counting triples using len(). I couldn't do that with the RDFa; I just cut and paste the HTML table into Excel and counted the number of rows. Fortunately every single file had exactly 700 triples, so I was good. Plus it seemed like a decent-enough test. In addition, I reviewed selections with my eyes, it looked good. So I think the results are good. But more robust tests, as Benjamin said, would be better.

**Comment by briesenberg07** [comment]
From 2023-11-14 UWLSWD meeting:

> - Validation vs. scripting to publish/output derivatives, etc…
>    - Keep these two separate?
>    - Or, as Theo says, implement validation in XML workflows… (‘schema-aware processing’)
> - Use XML tools for validation, or use DCTAP + RDF validation tools?? This is a question…
>    - OK, well, the things we want to validate… can these be expressed as DCTAP? As .xsd??
>    - QUESTIONS: Can DCTAP check lang tags? Can it check datatypes?? Can it handle ‘one of the above’
>    - See [DCTAP cookbook](https://dcmi.github.io/dctap/DCTAP-Cookbook.html) - it may hold some answers!!
>    - Is [tapshex](https://github.com/tombaker/tapshex) the DCTAP-to-ShEx tool we’ve been dreaming of??

**Comment by cspayne** [comment]
I've spent a bit of time looking at DCTAP, ShEx, and converting between them. 
[tapshex](https://github.com/tombaker/tapshex) is meant to convert DCTAP to ShEx, and it works to a degree; but to then validate data using that code, it needs a few adjustments and our DCTAP needs to be stricter, so I'm wondering whether this is a path we want to go down. 

Our [instances_of DCTAP](https://github.com/uwlib-cams/uwlswd/blob/main/dctap/instances_of.csv) will not produce workable ShEx that we can run using [PyShEx](https://github.com/hsolbrig/PyShEx/tree/master). 
I made some changes in a test file - [instances_of_test.csv](https://github.com/uwlib-cams/uwlswd/files/13708528/instances_of_test.csv) - and was able to produce the following ShEx which is _almost_ valid using [tap2shex on huggingface](https://huggingface.co/spaces/dcmi/dctap2shex): 

```
PREFIX dct: <http://purl.org/dc/terms/>
PREFIX schema: <http://schema.org/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
instancesOf  {
    rdf:type [<http://rdfs.org/ns/void#Dataset>];
    dct:title rdf:langString;
    dct:description rdf:langString;
    schema:disambiguatingDescription ["Linked Open Cultural Heritage Dataset"@en];
    dct:creator IRI;
    dct:publisher [<http://viaf.org/viaf/139541794>];
    dct:issued xsd:string;
    dct:language ["en"];
    schema:version xsd:string;
    dct:format IRI;
    dct:license IRI;
    dct:hasFormat IRI
}
```

In order to actually use this, 
- `start=@:instancesOf` needs to be added below the prefixes
- `:` needs to be added before `instancesOf`
- the schema iri needs to be corrected
- and `+` needs to be added after `dct:hasFormat IRI` because tap2shex doesn't account for the repeatable column. 

Also to note, if we are using DCTAP with conversion to ShEx in mind:
- valueDataType and valueConstraint cannot be used in the same row
- valueConstraint strings need to be in double quotes and valueConstraint IRIs need to be in <>
- we need to use IRI for our valueType not xsd:anyURI
- we need to use rdf:langString for string values with lang tags

**Reply by briesenberg07** [reply]
Thank you @cspayne ! I'd like to follow up on this further. 

If we're sure that 1) we've got 'valid' DCTAP and 2) Tap2shex is failing to output valid ShEx-C from this, I figure let's be good citizens and report the specific problems in the ShEx-C output. HuggingFace > [Tap2shex](https://huggingface.co/spaces/dcmi/dctap2shex) offers a way to start discussions or issues (at least, once I created an account).

I believe that you've already established both 1) and 2), but I'd like to catch up a bit before we report the issue. I'm going to work through the error corrections you provide for DCTAP/[instances_of.csv](https://github.com/uwlib-cams/uwlswd/blob/main/dctap/instances_of.csv) (thank you) and get it to the point (if I can) at which [dctap-python](https://github.com/dcmi/dctap-python) no longer gives me any error warnings. 

(Can't remember whether I mentioned [dctap-python](https://github.com/dcmi/dctap-python) to you @cspayne , but this is what I'm using to validate (not sure how strict this validation is...) DCTAP-as-DCTAP. It seems a bit clearer to me than using HuggingFace, where I don't see any specific options for validating the CSV itself. In dctap-python, on the other hand, I can run command...

```
> dctap read --warnings my_profile.csv
```

...to get a list of warnings re: the CSV itself. (Note that [example commands in the repo README are out-of-date](https://github.com/dcmi/dctap-python/issues/23), don't use `generate`, use `read`.)

**Reply by briesenberg07** [reply]
Hmm... one other note is that I'm not sure how strict the dctap-python validation really is. Warnings for my instances_of.csv (with none of the issues you found addressed) only yielded the following warnings:

```
WARNING [csv/column] Non-DCTAP element 'notes' not configured as extra element.
WARNING [instancesOf/shapeID] Value 'instancesOf' does not look like a URI.
WARNING [instancesOf/shapeID] Value 'instancesOf' does not look like a URI.
WARNING [instancesOf/shapeID] Value 'instancesOf' does not look like a URI.
WARNING [instancesOf/shapeID] Value 'instancesOf' does not look like a URI.
WARNING [instancesOf/shapeID] Value 'instancesOf' does not look like a URI.
WARNING [instancesOf/shapeID] Value 'instancesOf' does not look like a URI.
WARNING [instancesOf/shapeID] Value 'instancesOf' does not look like a URI.
WARNING [instancesOf/shapeID] Value 'instancesOf' does not look like a URI.
WARNING [instancesOf/shapeID] Value 'instancesOf' does not look like a URI.
WARNING [instancesOf/shapeID] Value 'instancesOf' does not look like a URI.
WARNING [instancesOf/shapeID] Value 'instancesOf' does not look like a URI.
WARNING [instancesOf/shapeID] Value 'instancesOf' does not look like a URI.
```

So, it seems to not be catching things? This seems like it'll make opening an issue or discussion more challenging--is the issue with the .CSV or the .CSV-to-ShEx conversion? etc.

**Reply by cspayne** [reply]
@briesenberg07 
If you change the shapeID from 'instancesOf' to ':instancesOf' with that ':' at the beginning, it will remove those errors. If I understand right, adding ':' is saying that this is a URI we are using to identify our shape. Currently we're only using this locally so (at least for now while we're just testing) it's okay that we haven't given it a namespace or an actual URI. 

Once that is done both instances_of.csv and instances_of_test.csv read as valid.

**Reply by cspayne** [reply]
@briesenberg07 
I found why this valueConstraint is showing up like this - we have it set as a picklist! 
>  Statement Template
            propertyID           schema:disambiguatingDescription
            propertyLabel        disambiguatingDescription
            repeatable           false
            valueDataType        xsd:string
            valueConstraint      ['Linked', 'Open', 'Cultural', 'Heritage', 'Dataset']
            valueConstraintType  **picklist**

When that is removed, it appears as a string. 

> Statement Template
            propertyID           schema:disambiguatingDescription
            propertyLabel        disambiguatingDescription
            repeatable           false
            valueNodeType        literal
            valueConstraint      "Linked Open Cultural Heritage Dataset"@en

**Reply by briesenberg07** [reply]
Thank you @cspayne !

Just pushed some corrections to [instances_of.csv](https://github.com/uwlib-cams/uwlswd/blob/main/dctap/instances_of.csv) from your feedback and ran CSV validation again using [dctap-python](https://github.com/dcmi/dctap-python).

I wanted to confirm that I could use picklists of literal values, picklists of IRIs, one-item picklists, and it seems that I can\*\*... see [20231220_dctap_warnings.txt](https://github.com/uwlib-cams/uwlswd/files/13731556/20231220_dctap_warnings.txt)

\*\* **_I jumped the gun somewhat here... I'm still a little murky on some of the results... for example:_**

```
Statement Template
            propertyID           schema:disambiguatingDescription
            propertyLabel        disambiguatingDescription
            valueConstraint      ['"Linked', 'Open', 'Cultural', 'Heritage', 'Dataset"@en']
            valueConstraintType  picklist
```
I don't think this is the desired result. I intended the full phrase to be the single item in the picklist, not each word
```
Statement Template
            propertyID           dct:publisher
            valueConstraint      ['<http://viaf.org/viaf/139541794>']
            valueConstraintType  picklist
            propertyID           dct:issued
            repeatable           false
            valueDataType        xsd:gYear
```
The quotes around my valueConstraint here are also confusing to me...

**Reply by briesenberg07** [reply]
One further question is whether it works to provide a prefixed IRI as a value constraint. I did this in the valueConstraint for rdf:type and it seems that it was interpreted as a literal instead :(

```
Statement Template
            propertyID           rdf:type
            propertyLabel        type
            repeatable           false
            valueConstraint      ['void:Dataset']
            valueConstraintType  picklist
```

**Reply by cspayne** [reply]
@briesenberg07 
I think some of this can be [configured](https://tapshex.readthedocs.io/en/latest/config/#) once a [config file is initialized](https://dctap-python.readthedocs.io/en/latest/cli/init/index.html#cli-subcommands-init), but I've been getting missing argument error when I try to run `dctap init`. 
Let me know if it works for you! If not, I think we should make an issue.

**Reply by briesenberg07** [reply]
Made some further edits to [instances_of.csv](https://github.com/uwlib-cams/uwlswd/blob/ea4679e8581d861a3839a4559be7445fa3cfa367/dctap/instances_of.csv) to try and eliminate the unexpected output described above. My remaining current DCTAP questions are, is `valueConstraintType : picklist` only for literal values? I probably just need to return to the DCTAP Primer and/or other docs. dctap-python output seems to me to interpret any picklist value as literal (encloses in quotes), see:

```
 Statement Template
            propertyID           dct:publisher
            propertyLabel        Publisher
            repeatable           false
            valueConstraint      ['<http://viaf.org/viaf/139541794>']
            valueConstraintType  picklist
...
Statement Template
            propertyID           dct:format
            propertyLabel        Format
            repeatable           false
            valueConstraint      ['<http://www.w3.org/ns/formats/JSON-LD>', '<http://www.w3.org/ns/formats/N-Triples>', '<http://www.w3.org/ns/formats/RDFa>', '<http://www.w3.org/ns/formats/RDF_XML>', '<http://www.w3.org/ns/formats/Turtle>']
            valueConstraintType  picklist
```
Removed `valueConstraintType : picklist` from the Statement Template for schema:disambiguatingDescription because it isn't needed here, but I still wonder how to provide a picklist of phrases... (see [comment above](https://github.com/uwlib-cams/uwlswd/discussions/66#discussioncomment-7911710)).

@cspayne I haven't yet looked at any of the information regarding config files for dctap-python, thank you for the links, I will try to run `dctap init` this week.

**Reply by cspayne** [reply]
Sounds good! A picklist of phrases should be doable if we can get a config file going - the picklist_item_separator is a space by default, but can be changed to a comma or something similar.

**Reply by cspayne** [reply]
From what I can tell by reading the docs, the two main issues I've found in regards to converting dctap to shexc are:
- ShEx has only one triple constraint for each property/triple, whereas DCTAP allows a valueNodeType, valueDataType, and valueConstraint
- DCTAP repeatable is not being mapped to ShEx +

**Reply by briesenberg07** [reply]
Interesting! Cardinality seems pretty important, surprising that this wouldn't be accounted for in the DCTAP-to-ShEx conversion. 

I still hope to put together some kind of reporting for the folks working on that. I believe my remaining to-do prior to that is to test creating a config file.
